azure/services/virtualmachines/spec.go

```diff
@@ -277,18 +291,77 @@ func (s *VMSpec) generateOSProfile() (*compute.OSProfile, error) {
 	return osProfile, nil
 }
 
-func (s *VMSpec) generateSecurityProfile() (*compute.SecurityProfile, error) {
+func (s *VMSpec) generateSecurityProfile(storageProfile *compute.StorageProfile) (*compute.SecurityProfile, error) {
 	if s.SecurityProfile == nil {
 		return nil, nil
 	}
 
-	if !s.SKU.HasCapability(resourceskus.EncryptionAtHost) {
-		return nil, azure.WithTerminalError(errors.Errorf("encryption at host is not supported for VM type %s", s.Size))
+	securityProfile := &compute.SecurityProfile{}
+
+	if storageProfile.OsDisk.ManagedDisk != nil &&
+		storageProfile.OsDisk.ManagedDisk.SecurityProfile != nil &&
+		storageProfile.OsDisk.ManagedDisk.SecurityProfile.SecurityEncryptionType != "" {
+		if s.SecurityProfile.EncryptionAtHost != nil && *s.SecurityProfile.EncryptionAtHost {
+			return nil, azure.WithTerminalError(errors.New("encryption at host is not supported when SecurityEncrytionType is set"))
+		}
+
+		securityProfile.UefiSettings = &compute.UefiSettings{}
+
+		if s.SecurityProfile.SecureBoot != "" && s.SecurityProfile.SecureBoot == infrav1.SecureBootPolicyEnabled {
+			securityProfile.UefiSettings.SecureBootEnabled = pointer.Bool(true)
+		} else {
+			securityProfile.UefiSettings.SecureBootEnabled = pointer.Bool(false)
+		}
+
+		if storageProfile.OsDisk.ManagedDisk.SecurityProfile.SecurityEncryptionType == compute.SecurityEncryptionTypesDiskWithVMGuestState && !*securityProfile.UefiSettings.SecureBootEnabled {
+			return nil, azure.WithTerminalError(errors.Errorf("secure boot should be enabled when SecurityEncrytionType is set to %s", compute.SecurityEncryptionTypesDiskWithVMGuestState))
+		}
+
+		if s.SecurityProfile.VirtualizedTrustedPlatformModule != "" && s.SecurityProfile.VirtualizedTrustedPlatformModule == infrav1.VirtualizedTrustedPlatformModulePolicyDisabled {
+			return nil, azure.WithTerminalError(errors.New("vTPM should be enabled when SecurityEncrytionType is set"))
+		}
+
+		securityProfile.UefiSettings.VTpmEnabled = pointer.Bool(true)
+
+		securityProfile.SecurityType = compute.SecurityTypesConfidentialVM
+
+		return securityProfile, nil
+	}
```

Comment:
> Would it be better if we enforce these validations at webhook level so that there are lesser chances for a user to run into Terminal Error ?

I absolutely agree, these validations are quite useful at the webhook level and we have them [here](https://github.com/kubernetes-sigs/cluster-api-provider-azure/pull/3265/files#diff-222a5b91649bbd8b784ec4d5476460390cc2f7e5839e63a968d719b3484ace68R395-R428). The latter is initially called through [here](https://github.com/kubernetes-sigs/cluster-api-provider-azure/blob/main/api/v1beta1/azuremachine_webhook.go#L57).  Does this makes sense?

Reply:
Makes sense, thanks! 

@CecileRobertMichon do you know why we perform validations in `generateSecurityProfile()` func? 
I am trying to understand the reasoning behind `generateSecurityProfile()` returning `azure.WithTerminalError()`. 
Can we not perform all the validations at webhook level ? Is it because we need `s` of type `*VMSpec` populated for these validations, and `s` can only be populated later in the reconciliation and not at the webhook level?